{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1c460704-e0f0-4433-ab54-697cf858f65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Declaring classes and importing \n",
    "### Coalescent Simulator Exercise\n",
    "\n",
    "import sys\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "# We are going to start by defining a few data structures (as classes) below.\n",
    "# This might seem confusing right now but this will make things much cleaner\n",
    "# below.\n",
    "class Node:\n",
    "    def __init__(self, descendants, nodeTime):\n",
    "        \"\"\"class method to create a new Node object.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        descendants -- a list of numbers corresponding to the individuals that descend from the new Node\n",
    "        nodeTime -- the time at which this node existed, expressed in generations ago (should be zero for present-day individuals)\n",
    "        \"\"\"\n",
    "        self.descendants = descendants\n",
    "        self.nodeTime = nodeTime\n",
    "\n",
    "class Branch:\n",
    "    def __init__(self, descendants, branchLength):\n",
    "        \"\"\"class method to create a new Branch object.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        descendants -- a list of numbers corresponding to the individuals that descend from the new Branch\n",
    "        branchLength -- the length of the branch, expressed in generations\n",
    "        \"\"\"\n",
    "        self.descendants = descendants\n",
    "        self.branchLength = branchLength\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "67cbaca4-5059-4756-9cc9-d6083ca720c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "\n",
    "def pickCoalTime(coalProb):\n",
    "    \"\"\"Incomplete function to pick coalescence time\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    coalProb -- the per-generation probability that a coalescence event will occur\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    coalTime -- the time at which the next coalescent event will occur\n",
    "    \n",
    "    \"\"\"\n",
    "    # Hint: What distribution describes how long we have to wait for an event to\n",
    "    # occur, given the probability that the event will occur in a single time-step?\n",
    "    # the answer is one of these: https://numpy.org/doc/stable/reference/random/generator.html\n",
    "   # raise NotImplementedError # remove this once you have finished the function\n",
    "    coalTime = np.random.default_rng().geometric(p=coalProb)\n",
    "    return coalTime "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "76c6e2ff-55dd-4951-ab11-d8ada82ed1bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def selectNodeIndicesToCoalesce(nodes):\n",
    "    \"\"\"Incomplete function to pick two node indices to coalesce\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    nodes -- the list of all nodes in the tree\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    nodeIndex1 -- the index of the first node in the next coalescence event\n",
    "    nodeIndex2 -- the index in the second node in the next coalescence event\n",
    "    \n",
    "    \"\"\"\n",
    "    # We need to figure out which two nodes to coalesce\n",
    "\n",
    "    indices = list(range(len(nodes)))\n",
    "    random.shuffle(indices)\n",
    "    nodeIndex1 = indices[0]\n",
    "    nodeIndex2 = indices[1]\n",
    " #   raise NotImplementedError # remove this once you have finished the function\n",
    "    return nodeIndex1, nodeIndex2\n",
    "\n",
    "### Tree-generating function will call your two functions above... ###\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f84a4237-9925-4cf2-b1bb-95a0dd2e2664",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your own code where you see triple hash marks \"###\"\n",
    "def generateTree(n, N):\n",
    "    \"\"\"Incomplete function to pick two node indices to coalesce\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    nodes -- the list of all nodes in the tree\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    nodes -- the list of all nodes that have yet to coalesce (should contain only one node)\n",
    "    branches -- the list of all branches in our completed coalescent tree\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # initialize our tree data structure\n",
    "    # we only need to keep track of two things for the purposes of this simulation:\n",
    "    #    1) for each node in the tree, a list of all individuals in our sample\n",
    "    #       that are descendants of the node\n",
    "    #    2) the time of this node (for our actual samples this is the present, or 0)\n",
    "\n",
    "    # initialize our tree nodes to contain our leaves\n",
    "    nodes = []\n",
    "    for j in range(n):\n",
    "        nodes.append(Node(descendants=[j], nodeTime=0))\n",
    "\n",
    "    # we will also keep track of all branches in the tree which will make it\n",
    "    # easier to throw down mutations randomly once we are done building our\n",
    "    # coalescent history.\n",
    "    branches = []\n",
    "\n",
    "    # we want to keep track of how far back in time we have gone\n",
    "    # we will use this variable to do that\n",
    "    currentTimeElapsed = 0\n",
    "\n",
    "    while len(nodes) >= 2:\n",
    "        i = len(nodes)\n",
    "        # calculate coalescence probability, ie. what is the per-generation\n",
    "        # probability that there will be any coalescence event among our lineages?\n",
    "        coalProb = (i*(i-1))/(4*N) ### note that i needs to be defined here!\n",
    "        \n",
    "        # Here is one of the ingenious things about coalescent simulation:\n",
    "        # we don't actually have to simulate each generation, and this is\n",
    "        # because the coalescent tells us what the distribution of\n",
    "        # coalescence times should be. All we have to do is draw from this\n",
    "        # distribution to get a number of generations, then we skip back\n",
    "        # to that time and continue our work. This makes the simulations\n",
    "        # really fast! We do this below:\n",
    "\n",
    "        # pick a random number of generations we will have to wait for the\n",
    "        # coalescence event to occur\n",
    "        coalGen = pickCoalTime(coalProb) ### this function needs to be finished above!\n",
    "\n",
    "        # pick a random pair of nodes from our tree that we will coalesce\n",
    "        nodeIndex1, nodeIndex2 = selectNodeIndicesToCoalesce(nodes)\n",
    "        node1, node2 = nodes[nodeIndex1], nodes[nodeIndex2]\n",
    "\n",
    "        # we just jumped back coalGen generations, so we will add this to our\n",
    "        # time counter\n",
    "        currentTimeElapsed += coalGen\n",
    "\n",
    "        # now that node1 and node2 have coalesced in our tree, we remove them from\n",
    "        # the set of nodes that have yet to coalesce:\n",
    "        nodes.remove(node1)\n",
    "        nodes.remove(node2)\n",
    "\n",
    "        # but now we must replace those nodes with their ancestor's node:\n",
    "        mergeDescendants = node1.descendants + node2.descendants\n",
    "        nodes.append(Node(descendants=mergeDescendants, nodeTime=currentTimeElapsed)) # replace the blahs here with something useful!\n",
    "\n",
    "        # in essence we have added two new branches to the tree: one from node1\n",
    "        # and one from node2; these two branches connect to each other at\n",
    "        # currentTimeElapsed, and their descendents\n",
    "        # Note the benefit of using the Node class here: we don't have to remember the internals\n",
    "        # of how the Node data structure works. We just need to know that it needs a list of\n",
    "        # descendants and a time and our class takes care of the rest. For example, we could have\n",
    "        # simply created a list of two elements (the first being itself a list of descendants,\n",
    "        # and the second being a number corresponding to the time), but then we would always have\n",
    "        # to remember that node[0] has node's descendants and node[1] contains the time. That\n",
    "        # code would also be harder read!\n",
    "\n",
    "        # eventually we will be throwing mutations down onto branches;\n",
    "        # to do this we need to keep track of two things:\n",
    "        #     1) the length of the branch (in number of generations)\n",
    "        #     2) the set of nodes that are a descendant of this branch\n",
    "        # the latter will tell us which individuals will harbor any mutation(s)\n",
    "        # that will occur on this branch\n",
    "        # first, let's get the length of the branch coming from node1:\n",
    "        branchLength1 = currentTimeElapsed - node1.nodeTime\n",
    "\n",
    "        # let's record this length along with all individuals node1 is \n",
    "        # ancestral to, which we have been keeping track of all along!\n",
    "        branches.append(Branch(descendants=node1.descendants, branchLength=branchLength1))\n",
    "        # Once again our use of classes has simplified things for us.\n",
    "\n",
    "        # do the same thing for the branch leading to node2\n",
    "        branchLength2 = currentTimeElapsed - node2.nodeTime\n",
    "        branches.append(Branch(descendants=node2.descendants, branchLength=branchLength2))\n",
    "    return nodes, branches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "bbe44247-1801-4b19-9936-5d2279966976",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTotalTreeHeight(nodes):\n",
    "    \"\"\"Missing function to get the height of our tree\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    ??? -- ???\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    tmrca -- the time back to the common ancestor of our complete sample\n",
    "    \n",
    "    \"\"\"\n",
    "    tmrca = nodes[0].nodeTime\n",
    "    \n",
    "    # Hint: the time to the most recent common ancestor is simply the age of the final\n",
    "    # node that we added to the tree\n",
    "    return tmrca\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b2f8cfd7-188b-4e90-ae89-342cc81d3235",
   "metadata": {},
   "outputs": [],
   "source": [
    "def drawNumberOfMutationsOnBranch(branchLengthInGen, totalMutationRatePerGen):\n",
    "    \"\"\"Missing function to get the number of mutations we want to place on a given branch\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    branchLengthInGen -- The length of the current branch, expressed in generations\n",
    "    totalMutationRatePerGen -- The mutation rate per generation across our entire L-base pair locus\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    numMuts -- the number of mutations\n",
    "    \n",
    "    \"\"\"\n",
    "   # totalMutationRatePerGen = mu * L\n",
    "   # totalBranchLengthInGen = 1000\n",
    "    rng = np.random.default_rng()\n",
    "    numMuts = rng.binomial(branchLengthInGen, totalMutationRatePerGen)\n",
    "    # have to pick an appropriately distributed random number of mutations for\n",
    "    # this branch.\n",
    "   \n",
    "    return numMuts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "269d326a-cf67-4713-81f4-406c80a362fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def throwDownMutations(branches, mu, L):\n",
    "    \"\"\"Missing function to get the number of mutations we want to place on a given branch\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    branches -- Our tree (represented by a list of branches)\n",
    "    mu -- The per-base pair mutation rate\n",
    "    L -- the total number of base pairs in our simulated region\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    numMuts -- the number of mutations\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # First, initialize our matrix to n empty lists\n",
    "    \n",
    "    seqMatrix = []\n",
    "    for i in range(n):\n",
    "        seqMatrix.append([])\n",
    "\n",
    "    # Now we are ready to throw down our mutations. There are a few ways that we could\n",
    "    # go about it, but the way I have written it below is to iterate through each\n",
    "    # branch of the tree, pick an appropriate random number of mutations for that branch\n",
    "    # (you have already coded that part), and then for each chromosome in our sample write a 1\n",
    "    # if it is a descendant of the branch where the mutation occurred and 0 otherwise.\n",
    "    for branch in branches:\n",
    "        numMuts = drawNumberOfMutationsOnBranch(branch.branchLength, mu*L)\n",
    "        for mut in range(numMuts):\n",
    "            for i in range(n):\n",
    "                if i in branch.descendants:\n",
    "                    seqMatrix[i].append(1)\n",
    "                else:\n",
    "                    seqMatrix[i].append(0)\n",
    "    return seqMatrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "860845c6-a5f5-40c4-b774-19e6c77c388f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Coalescent Simulation Parameters -- you can mess with these a little if you want\n",
    "# but if you make L and N too huge things might get slow. Your results might also \n",
    "# start to act a little weird if you make n very large.\n",
    "L = 10000  # Locus length\n",
    "N = 1e4  # Population Size\n",
    "n = 4  # number of pairs of genes that can coalesce\n",
    "mu = 1.2e-8  # per site mutation rate\n",
    "\n",
    "# Simulate\n",
    "nodes, branches = generateTree(n, N)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0b6d20c9-106c-455e-84c8-7f0c7add6709",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "observed time to most recent common ancestor: 29242\n",
      "expected time to most recent common ancestor: 30000.0\n",
      "00011\n",
      "00100\n",
      "11100\n",
      "11100\n"
     ]
    }
   ],
   "source": [
    "tmrca = getTotalTreeHeight(nodes)\n",
    "print(\"observed time to most recent common ancestor:\", tmrca) # if you do this a bunch of times, you should get a result equal to the following:\n",
    "eTMRCA = 0\n",
    "for i in range(2, n+1):\n",
    "    eTMRCA += (4*N)/(i*(i-1))\n",
    "print(\"expected time to most recent common ancestor:\", eTMRCA)\n",
    "# Are you getting the (approximately) right answer?\n",
    "\n",
    "seqMatrix = throwDownMutations(branches, mu, L)\n",
    "for seq in seqMatrix:\n",
    "    print(\"\".join([str(x) for x in seq]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a109c762-124c-44c1-974f-c695b21c0e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Optional: if you would like, you can also compute the site frequency spectrum and calculate pi\n",
    "\n",
    "### Your Analysis here:\n",
    "# Site frequency Spectrum\n",
    "# Come up with a way to generate the SFS from the simulated seqMatrix we generated above\n",
    "#\n",
    "# You could also plot it if you like:\n",
    "# import matplotlib.pyplot as plt\n",
    "# plt.hist() # will help make the plot (in fact, you can use this to calculate the SFS and plot it all in one shot!\n",
    "# plt.savefig(\"sfs.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d3c9521-7ccd-4f2b-b25a-2f0f2669dd01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# partially completed function for calculating Pi\n",
    "def pi(seqMatrix, L):\n",
    "    \"\"\"Missing function to get the number of mutations we want to place on a given branch\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seqMatrix -- A matrix of genotypes\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    pi -- the expected heterozygosity of our sample, estimated from seqMatrix\n",
    "    \n",
    "    \"\"\"\n",
    "    raise NotImplementedError # remove this once you have finished the function\n",
    "    ### hint: j is a numpy array of derived allele counts for each SNP\n",
    "    ### hint: n is sample size\n",
    "\n",
    "    pi = sum((2*j*(n-j))/(n*(n-1)))\n",
    "    return pi/L\n",
    "\n",
    "print(\"pi in sample:\", pi(seqMatrix, L))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
